### PR TITLE
[Core] Add specification/trait C++ implementation

### PR DESCRIPTION
**Pull request by feltech**
For #250 we require the C++ infrastructure to support specifications following the new design of #238 et al. 

This PR introduces an initial draft of specifications as dictionary-like data structures and traits as views on a specification, coded in C++ with Python bindings, and tested exclusively through the Python bindings.

Initial performance tests using #270 and reported in #253 lead us to believe the performance of this dictionary-like design is unlikely to be a serious bottleneck.

The Doxygen config is updated to render docs for the new C++ classes.

### CODE REVIEW

File: src/openassetio-core/include/openassetio/specifications.hpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2013-2022 The Foundry Visionmongers Ltd
+#pragma once
+
+#include <memory>
+#include <string>
+#include <unordered_map>
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/specifications.hpp:7]: Capturing off-line discussion:

Work out strategy to make sure we don't leak the storage implementation.

File: src/openassetio-core/include/openassetio/specifications.hpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+#include <utility>
+#include <variant>
+#include <vector>
+// API export generated header
+#include <openassetio/export.h>
+
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/specifications.hpp:13]: [fix] Add High-level architectural manifesto of what is in this file before too much implementation happens 📜

File: src/openassetio-core/include/openassetio/specification/Specification.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+class OPENASSETIO_CORE_EXPORT Specification {
+ public:
+  using TraitIds = std::vector<trait::TraitId>;
+
+  explicit Specification(const TraitIds& traitIds);
+  virtual ~Specification();
+
+  [[nodiscard]] bool hasTrait(const trait::TraitId& traitId) const;
+
+  [[nodiscard]] bool getTraitProperty(trait::property::Value* out, const trait::TraitId& traitId,
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/specification/Specification.hpp:30]: [question] Do we want to consider typed accessors here?

File: src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+ protected:
+  SpecificationPtr& spec() { return spec_; }
+  [[nodiscard]] const SpecificationPtr& spec() const { return spec_; }
+
+  template <class T>
+  [[nodiscard]] bool getTraitProperty(T* out, const TraitId& traitId,
+                                      const property::Key& propertyKey) const {
+    if (property::Value value; spec()->getTraitProperty(&value, traitId, propertyKey)) {
+      // TODO(DF): Make exception message more friendly.
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp:39]: [question] Do we want to `return false` instead of excepting, if the goal of the bool/out param was to avoid exceptions in this API? Otherwise, you need both wrappers for most typical get scenarios, as missing or mistyped are potentially equally likely?

**Review by foundrytom — COMMENTED** [review]
WIP, but some initial Qs

File: python/openassetio/specification/__init__.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+"""
+Entity type system and property transport mechanism.
+"""
+__all__ = ["BlobSpecification", "Specification"]
+
+from openassetio._openassetio.specification import BlobSpecification, Specification
```
[the remark this answers is not in the document]
Comment (feltech) [reply python/openassetio/specification/__init__.py:21]: This came from an annoyance with my IDE, which is also now causing problems with `pylint` for a similar reason.  

The `python/openassetio` source directory doesn't contain `_openassetio.so` (since it's installed into the venv), so relative imports fail, tripping up my IDE code completion.  I attempted to work around this by making it an absolute import.  So I made it absolute just for selfish reasons and forgot about it.  

However, now with `pylint` the same problem occurs. Linting fails because of unrecognized imports, even if it's executed from the venv where `_openassetio.so` is installed.  This is apparently because the source vs. installed `openassetio` package names conflict (as evidenced by the fact that `pylint` on the `test` directory succeeds).

Placing `_openassetio.so` alongside, rather than within, the `openassetio` package would make dev life much easier on several fronts.  E.g. we may be able to re-introduce `--editable` installs for running tests, speeding up the TDD cycle.

I understand that we don't want to structure our package based on what's convenient for dev, though. 

We could run `pylint` on the installed package, working around that problem, but IDE code completion would still suffer.

We could "install" the Python extension module to the source directory, so that relative imports work in IDE/pylint, then add it as an installable artefact in `setup.py`, which may be the correct thing to do anyway (I don't know much about `setup.py` and extension modules).

[the remark this answers is not in the document]
Comment (feltech) [reply python/openassetio/specification/__init__.py:21]: In the meantime we can switch to absolute imports and disable the pylint checks as a quick fix rather than worrying about it in this PR.  But I would really like to solve this problem sooner rather than later - having good code completion (and thorough linting) is extremely useful.

File: src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp (hunk 4 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp:39]: I'm not sure about this. I didn't view the `bool` as a way to avoid all exceptions, but rather that "unset" is not an exceptional case, and is certainly not an error.

Conversely, I think a bad `T` is an exceptional error case, especially as users should be using well-behaved pre-agreed trait wrappers rather than the low-level arbitrary-dict API.  

I'm also not sure that it's a good idea to conflate "not set" with "set, but with an unexpected value type", with no way to know which.

But maybe there is a counterexample where such a case is valid and not an error?

File: src/openassetio-core/include/openassetio/specification/Specification.hpp (hunk 3 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:30]: We can move the typed accessor from `TraitBase` to here.  I was in two minds where it belongs. 

I guess I viewed `getTraitProperty` as an advanced method that normal users would never call, since they should be using trait views, so didn't want to be too friendly.

The wrapper typed accessor in `TraitBase` is also `protected`, reducing the public API surface area, whereas if we move it here it'll have to be `public`.

Do you think it's worth adding as part of the public interface?

File: python/openassetio/specification/trait/__init__.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+"""
+Trait classes providing a concrete view on specification properties.
+"""
+__all__ = ["BlobTrait"]
+
+from openassetio._openassetio.specification.trait import BlobTrait
```
[the remark this answers is not in the document]
Comment (feltech) [reply python/openassetio/specification/trait/__init__.py:21]: I don't have a strong opinion on this either way.  I thought in terms of specifications composing traits, so a parent-child kind-of thing.  But as you say, that's not really true in practice.  Maybe a peer relationship does make more sense.

File: python/openassetio/specification/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (foundrytom) [reply python/openassetio/specification/__init__.py:21]: Oh man, that sounds like a _load_ of fun. Lets scheme on this sooner rather than later.

File: python/openassetio/specification/trait/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (foundrytom) [reply python/openassetio/specification/trait/__init__.py:21]: I think I leant towards peer on the basis of `managementPolicy` and `entityTraits`

File: src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp (hunk 4 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp:39]: Jotting down an off-line discussion:

The main thing we need to figure out is that is incorrect type an "exceptional" case, or does its very possibility force consumers to always defensively code around it, mitigating some of the semantic value brought by moving `InvalidKey` to the bool-outptr pattern.

File: src/openassetio-core/include/openassetio/specification/Specification.hpp (hunk 3 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:30]: Capturing off-line discussion:

For now, the main choice is: do we leak `std::variant` in the public API?

File: src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp (hunk 4 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp:39]: Related to this: the current implementation will also throw if `traitId` doesn't exist in the dict (i.e. the specification doesn't support the requested trait).

File: src/openassetio-core/include/openassetio/specifications.hpp (hunk 2 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/specifications.hpp]: Many things changed, but somewhat related, added some docstrings in 6a50c5f9

File: src/openassetio-core/include/openassetio/specifications.hpp (hunk 1 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/specifications.hpp]: Used pimpl in the rewrite in 01d6a8ea

File: python/openassetio/specification/trait/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (feltech) [reply python/openassetio/specification/trait/__init__.py]: Moved `trait` outside of `specification` in 831a9168.

**Review by foundrytom — COMMENTED** [review]
A few more open questions

File: src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+struct TraitBase {
+  using SpecificationPtr = std::shared_ptr<Specification>;
+
+  explicit TraitBase(SpecificationPtr spec) : spec_{std::move(spec)} {}
+
+  [[nodiscard]] static const TraitId& traitId() { return Derived::kId; }
+  [[nodiscard]] bool isValid() const { return spec_->hasTrait(traitId()); }
+
+ protected:
+  SpecificationPtr& spec() { return spec_; }
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp]: In general I'd err on the side of being verbose.  In this case I thought it was pretty unambiguous.  But I'll happily expand it.

File: tests/openassetio/test_traits.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    def test_when_trait_is_not_found_then_IndexError_raised(self, a_specification):
+        with pytest.raises(IndexError):
+            a_specification.setTraitProperty("unknown_trait", "a string", "string")
+
+        with pytest.raises(IndexError):
+            _ = a_specification.getTraitProperty("unknown_trait", "a string")
+
+    def test_when_value_is_not_supported_then_set_raises_TypeError(self, a_specification):
+        with pytest.raises(TypeError):
+            a_specification.setTraitProperty("first_trait", "unknown type", object())
```
[the remark this answers is not in the document]
Comment (feltech) [reply tests/openassetio/test_traits.py:54]: Yeah, I just didn't want to splat any existing code (yet), whilst keeping the new stuff separate.

File: src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp (hunk 4 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp]: Capturing offline discussion:

The choice comes from the fact that we had previously made an API-level decision that "missing" is less important than "invalid type". Upon reflection, we can't say that for sure. Without doubt, invalid type is always bad, but in many cases missing is equally so.  Having two mechanism to categorise the failure reason is confusing for everyone, and forces an opinion on what is more or less important.

Proposal:

- `C`/`C++` Core API:  Return an `enum` instead of a `bool` to cover `ok`, `not_set` and `invalid_type`.
- Future `C++` extension: return `optional<variant<>>` to allow the host to use more standard techniques to determine the status.
- `Python`: Return `None` or value by default, add `strict` mode that throws exceptions if missing/invalid type

File: src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp]: Verbose is good from my p.o.v too as an OSS project, if you're happy.

File: src/openassetio-core/include/openassetio/specifications.hpp (hunk 2 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specifications.hpp]: These are a good start. I guess the main request was for something more along the lines of what even is a spec, when is used, the whole view/data separation and relation to traits.

File: tests/openassetio/test_traits.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+        actual = trait.BlobTrait(a_blob_specification).getUrl()
+
+        assert actual == expected
+
+
+class Test_BlobTrait_setURL:
+    def test_when_spec_doesnt_have_blob_trait_then_raises_IndexError(self, a_specification):
+        with pytest.raises(IndexError):
+            trait.BlobTrait(a_specification).setUrl("some://url")
```
[the remark this answers is not in the document]
Comment (feltech) [reply tests/openassetio/test_traits.py:100]: Does it then make sense to have concrete specification classes, i.e. `BlobSpecification`?  Since you could transform it into a different specification by setting new traits on it.

[the remark this answers is not in the document]
Comment (foundrytom) [reply tests/openassetio/test_traits.py:100]: It felt like the concrete specs are more generally a template for consistency. The reason for an `ImageSpecification` being to ensure that all hosts.managers dealing with an image agree it combines `blob`, `image`, `raster` for example, and make it easy for a programmer to know what should/shouldnt be available (as there aren't trait accessors on that spec for traits an "image" doesn't have).

I guess it throws up the Q as to whether:
 - Direct trait use is allowed (feels like it should be) take `register` for example, wanting to just use a `BlobTrait`, and we dont need `BlobSpecification` specifically unless its composes multiple traits.
 - Or, should only `Specifications` be used, and we consequently always have a spec for each trait.

[the remark this answers is not in the document]
Comment (foundrytom) [reply tests/openassetio/test_traits.py:100]: I guess, coupled with the other comment above, is this desirable for a manager in resolve:

```python

result = Specification()

if myAsset.filePath:
  blob = BlobTrait(result)
  blob.setUrl(f"file://{myAsset.filePath}")

if myAsset.isSequence:
  range = FrameRangeTrait(result)
  range.setStartFrame(myAsset.startFrame)
...
```

[the remark this answers is not in the document]
Comment (foundrytom) [reply tests/openassetio/test_traits.py:100]: .. though it does pose the question of "does just wrapping add the trait ID", maybe we should look at how applied schemas work in USD.

File: src/openassetio-core/include/openassetio/specification/Specification.hpp [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+   *
+   * @param traitId ID of trait to update.
+   * @param propertyKey Key of property to set.
+   * @param propertyValue Value to set.
+   * @exception `std::out_of_range` if the trait is not supported by
+   * this specification.
+   */
+  void setTraitProperty(const trait::TraitId& traitId, const trait::property::Key& propertyKey,
+                        trait::property::Value propertyValue);
+
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/specification/Specification.hpp:81]: [question] Are we considering introspection one for a follow up PR? Ie, the ability to iterate traits and their properties as for serialisation.

Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:81]: I think it's probably a good idea to consider serialisation at a later date.  

Serialisation is not really needed for the the minimal "manager that refuses to manage anything" vertical slice.  We could define a new vertical slice that requires serialisation, with properly designed workflows and end-to-end tests, as a new unit of work?

File: tests/openassetio/test_traits.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (feltech) [reply tests/openassetio/test_traits.py:100]: From what I read, USD's applied API schemas do indeed add the attributes. Non-applied schemas (which are similar to how I have viewed traits so far) do not.  So, as in all things, USD has it both ways.  I'm not 100% that's how applied vs. non-applied schemas work, but that's my understanding.

File: src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp (hunk 4 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp]: Updated in b1169a1a

File: tests/openassetio/test_traits.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+        assert a_specification.getTraitProperty("first_trait", "a string") == "string"
+        assert isinstance(a_specification.getTraitProperty("first_trait", "a string"), str)
+        assert a_specification.getTraitProperty("second_trait", "an int") == 1
+        assert isinstance(a_specification.getTraitProperty("second_trait", "an int"), int)
+        assert a_specification.getTraitProperty("first_trait", "a float") == 1.0
+        assert isinstance(a_specification.getTraitProperty("first_trait", "a float"), float)
+        assert a_specification.getTraitProperty("second_trait", "a bool") is True
+
+    def test_when_key_is_not_found_then_get_returns_None(self, a_specification):
```
[the remark this answers is not in the document]
Comment (feltech) [reply tests/openassetio/test_traits.py:42]: Ah, hmm, yeah I have used "missing" vs. "found", e.g. in the enum added in b1169a1a.  Would you prefer "set" vs "unset"?  My slight reservation is that these are verbs as well as adjectives.

File: tests/openassetio/test_traits.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        assert a_specification.getTraitProperty("first_trait", "a string") is None
+
+    def test_when_trait_is_not_found_then_IndexError_raised(self, a_specification):
+        with pytest.raises(IndexError):
+            a_specification.setTraitProperty("unknown_trait", "a string", "string")
+
+        with pytest.raises(IndexError):
+            _ = a_specification.getTraitProperty("unknown_trait", "a string")
+
+    def test_when_value_is_not_supported_then_set_raises_TypeError(self, a_specification):
```
[the remark this answers is not in the document]
Comment (feltech) [reply tests/openassetio/test_traits.py:52]: Really good spot.  Adding a test for `None` revealed rather unintentional behaviour (was silently converted to `false`). Fixed in 6e68a638

File: src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @tparam Derived Concrete subclass.
+ */
+template <class Derived>
+struct TraitBase {
+  using SpecificationPtr = std::shared_ptr<Specification>;
+
+  explicit TraitBase(SpecificationPtr spec) : spec_{std::move(spec)} {}
+
+  [[nodiscard]] static const TraitId& traitId() { return Derived::kId; }
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp]: Renamed in 996495a3

File: src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/trait/TraitBase.hpp]: Renamed in 8e17d811

File: src/openassetio-core/include/openassetio/specification/Specification.hpp (hunk 10 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:81]: Sounds good to me 👍

File: src/openassetio-core/include/openassetio/specification/Specification.hpp (hunk 3 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:67]: We can certainly save this for a later PR if we're happy that we're leaking `std::variant` in the public API.

File: tests/openassetio/test_traits.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (foundrytom) [reply tests/openassetio/test_traits.py:42]: I think the verb/adjective ambiguity is a great point.

File: tests/openassetio/test_traits.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+        assert a_specification.getTraitProperty("second_trait", "an int") == 1
+        assert isinstance(a_specification.getTraitProperty("second_trait", "an int"), int)
+        assert a_specification.getTraitProperty("first_trait", "a float") == 1.0
+        assert isinstance(a_specification.getTraitProperty("first_trait", "a float"), float)
+        assert a_specification.getTraitProperty("second_trait", "a bool") is True
+
+    def test_when_key_is_not_found_then_get_returns_None(self, a_specification):
+        assert a_specification.getTraitProperty("first_trait", "a string") is None
+
+    def test_when_trait_is_not_found_then_IndexError_raised(self, a_specification):
```
[the remark this answers is not in the document]
Comment (foundrytom) [reply tests/openassetio/test_traits.py:45]: Yeah, not required for this PR

File: tests/openassetio/test_traits.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (foundrytom) [reply tests/openassetio/test_traits.py:122]: We can re-visit adding traits later, but definitely need to make sure that we highlight the use of `isValid` in the docs.

File: src/openassetio-core/include/openassetio/trait/TraitBase.hpp [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+   * @param specification Underlying specification to wrap.
+   */
+  explicit TraitBase(SpecificationPtr specification) : specification_{std::move(specification)} {}
+
+  /**
+   * Get the ID of this trait.
+   *
+   * @return ID of this trait.
+   */
+  [[nodiscard]] static const TraitId& id() { return Derived::kId; }
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/trait/TraitBase.hpp:39]: [fix] Capturing an off-line conversation:

If we don't need to pass around base class `TraitPtr`s, then do we need this method? Is `kId` sufficient?

File: src/openassetio-core/include/openassetio/trait/BlobTrait.hpp [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+ * Core trait class representing a locatable blob of data of a specific
+ * type.
+ *
+ * Has the ID of "blob" and defines "url" and "mimeType" properties.
+ */
+struct OPENASSETIO_CORE_EXPORT BlobTrait : TraitBase<BlobTrait> {
+  /// ID of this trait.
+  static inline const TraitId kId{"blob"};
+  /// Key for URL property.
+  static inline const property::Key kUrl{"url"};
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/trait/BlobTrait.hpp:28]: [consider] Should we prefix property key names to avoid ambiguity? eg `kProperty_url` or similar, so that it is clear that `kId` isn't a property called `id`?

**Review by foundrytom — CHANGES_REQUESTED** [review]
Adding a few off-line chats.

File: src/openassetio-core/include/openassetio/specification/BlobSpecification.hpp [hunk 17]
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/BlobSpecification.hpp]: Removed in dfb36cd7

File: src/openassetio-core/include/openassetio/trait/BlobTrait.hpp (hunk 16 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/trait/BlobTrait.hpp]: I went a slightly different direction in 49c7861c, where I made `kUrl` and `kMimeType` private implementation details.  My argument is that if we do keep them as public members, then we should also expose them to Python.  So, since YAGNI I went with just making them private.

File: src/openassetio-python/trait/BlobTrait.cpp [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+#include <optional>
+#include <type_traits>
+
+#include <pybind11/stl.h>
+
+// Public.
+#include <openassetio/export.h>
+#include <openassetio/trait/BlobTrait.hpp>
+// Private.
+#include <trait/keys.hpp>
```
Comment (feltech) [inline src/openassetio-python/trait/BlobTrait.cpp:12]: [question] Here we're `#include`ing a header that is not part of the public API, i.e. not in the `include` directory.  

To make this work, the build target's include directories must overlap, i.e. both the `openassetio-core` directory and the `openassetio-core/include` directory are on the search path.  

I think the directory structure should be rejigged to remove the overlap, but I can't decide a good split.

Currently we have
```
openassetio-core
├── include
│   └── openassetio
│       ├── managerAPI
│       ├── specification
│       └── trait
├── managerAPI
├── specification
└── trait
```

We could e.g. restructure to something like
```
openassetio-core
├── include
│   └── openassetio
│       ├── managerAPI
│       ├── specification
│       └── trait
└── src
    └── private
        ├── managerAPI
        ├── specification
        └── trait
```
which would allow `#include <private/trait/keys.hpp>` (assuming `src` was on the include search path).  

I don't particularly like this option, but it's surprisingly hard to come up with a a scheme that doesn't overlap, isn't overly verbose, gives clear at-a-glance `#include`s, whilst (ideally) dog-fooding our use of public includes.

We can look at other projects to get some inspiration.

Comment (foundrytom) [reply src/openassetio-python/trait/BlobTrait.cpp:12]: How come its not part of the public API? If it's needed for a custom trait, then it needs to be public - as traits have to be runtime extensible?

File: src/openassetio-core/include/openassetio/trait/BlobTrait.hpp (hunk 16 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/trait/BlobTrait.hpp]: Should these be private to the implementation of `BlobTrait`, not to the API? _Any_ trait must be considered out-of-core so we can't have anything shared/etc.. that can't be done in a totally random, unconnected repo with just a copy of the source.

Comment (foundrytom) [reply src/openassetio-core/include/openassetio/trait/BlobTrait.hpp]: Also, was the removal of these to force the use of a traits as a view on a `Specification` instance, instead of using their key constants directly in `Specification::getTraitProperty`? Though higher overhead, that's probably a 'good thing' as there have been comments before that more than one way to do the same thing causes confusion.

File: src/openassetio-python/trait/BlobTrait.cpp (hunk 18 above)
Comment (foundrytom) [reply src/openassetio-python/trait/BlobTrait.cpp:12]: I think I just answered that Q when looking into the comment above (sorry was visiting these in reverse order), either way, we should probably have a huddle on src layout?

File: src/openassetio-core/include/openassetio/trait/BlobTrait.hpp (hunk 16 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/trait/BlobTrait.hpp]: The only reason they were public was so the Python bindings could use them in crafting an exception message, which seemed like a bad excuse for making something public and, like you say, unintentionally encouraging their use as keys in a raw `getTraitProperty` query.

Good point that we want to keep the traits self-contained, though.  I can rename the `keys.hpp` to `BlobTrait.hpp` so it's clear.

Comment (feltech) [reply src/openassetio-core/include/openassetio/trait/BlobTrait.hpp]: Updated in 40cf7353

File: src/openassetio-core/include/openassetio/specifications.hpp (hunk 2 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/specifications.hpp]: Did a second pass in 55fbc07a. Also enabled Doxygen for public C++ headers in c905c47d.

Perhaps still need a little fleshing out - more example snippets, perhaps.

That reminds me, one cool thing I've done with Doxygen in the long lost past is [snippets](https://www.doxygen.nl/manual/commands.html#cmdsnippet), where you can mark a section of source code with special comments, then refer to that section in the docs, which will reproduce it.  So you have a living example.  Would be great to eventually have some "workflow test" snippets that we can litter through the docs.

Comment (feltech) [reply src/openassetio-core/include/openassetio/specifications.hpp]: Restructured a bit in d5a7251b - how's it looking re. this review comment?

File: python/openassetio/specification/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (feltech) [reply python/openassetio/specification/__init__.py]: I disabled the pylint checks back in 4348ffb0 - though none of the Github checks are running at the moment for some reason :thinking:

File: src/openassetio-core/include/openassetio/specification/Specification.hpp (hunk 3 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:106]: Do we need to chat about this some more?  It sounds like we do want to expose `std::variant`, along with (eventually) a `std::optional` API, to make the modern C++ folks happy?

But if you think we should hide it for now we can do that.  Hiding `std::variant` makes the Python bindings a bit trickier - pybind is clever enough to work with `std::variant` out of the box, so if we hide it's use behind a typed accessor then we'll need a way to "unhide" it (or equivalent) in the bindings.

Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:106]: One argument perhaps is that it's just coincidental that we return a `std::variant` as well as use them internally.  We can have an API that returns `std::variant` and swap out the underlying storage for something else, they're not necessarily coupled.

File: src/openassetio-core/include/openassetio/trait/BlobTrait.hpp (hunk 16 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/trait/BlobTrait.hpp]: ... and then removed in 077c393f after off-line discussion, where we decided that simplifying the exceptions is a small price to pay for dodging this question entirely ;)

File: src/openassetio-core/include/openassetio/specification/Specification.hpp [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+   * @param[out] out Storage fo result, only written to if the property
+   * is set.
+   * @param traitId ID of trait to query.
+   * @param propertyKey Key of trait's property to query.
+   * @return `true` if value was found, `false` if it is unset.
+   * @exception `std::out_of_range` if the trait is not supported by
+   * this specification.
+   */
+  [[nodiscard]] bool getTraitProperty(trait::property::Value* out, const trait::TraitId& traitId,
+                                      const trait::property::Key& propertyKey) const;
```
Comment (foundrytom) [inline src/openassetio-core/include/openassetio/specification/Specification.hpp:107]: [fix] We should probably also document what constitutes a valid trait name/property key

File: src/openassetio-core/include/openassetio/specification/Specification.hpp (hunk 3 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:106]: Shall we leave it as a variant for now, and pick this up in another ticket in the not too distant?

**Review by foundrytom — COMMENTED** [review]
Sorry I was sure I posted these yesterday, maybe it got lost in the GitHub glitch

File: src/openassetio-core/include/openassetio/specifications.hpp (hunk 2 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specifications.hpp]: That is spot on!

File: src/openassetio-core/include/openassetio/specification/Specification.hpp (hunk 19 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:107]: I'm not sure what you mean, as in, it's any arbitrary string, as can be found by looking at the type.  Do we have further restrictions to add?  I guess the trait ID needs to be globally unique (I think I've called that out somewhere).

Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:107]: Yeah, was wondering if its fine for them to contain any valid UTF-8 character, and how that then maps to method names, etc.

File: src/openassetio-core/include/openassetio/trait/BlobTrait.hpp [hunk 20]
```diff
@@ -20,6 +20,20 @@ namespace trait {
  * type.
  *
  * Has the ID of "blob" and defines "url" and "mimeType" properties.
+ *
+ * This core trait defines one of the most common set of properties in
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/trait/BlobTrait.hpp:24]: We discussed off-line and decided to leave it in for now, but likely remove in favour of moving it to the [OpenAssetIO-MediaCreation](https://github.com/TheFoundryVisionmongers/OpenAssetIO-MediaCreation) repo

File: src/openassetio-core/include/openassetio/trait/TraitBase.hpp [hunk 21]
```diff
@@ -17,6 +17,53 @@ namespace trait {
 /**
  * Abstract CRTP base class for specification traits.
  *
+ * This non-virtual templated base class provides the common interface
+ * for a concrete trait view on a specification::Specification.
+ *
+ * A trait view provides a way to hide the underlying dictionary-like
+ * data access from hosts and managers.
```
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/trait/TraitBase.hpp]: Restructured slightly in bc850855 and fixed the example.

[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/trait/TraitBase.hpp]: At the risk of dragging this out further, the example in the docs is rather boilerplatey, reproducing it here:
```c++
property::Int myValue = 123;  // Default if property not found.

MyTrait myTrait{specification};

if (myTrait.isValid()) {

  if (myTrait.getMyValue(&myValue) != TraitPropertyStatus::kFound) {

    std::cerr << "Warning: myValue not found\n";
  }
}
```
We could simplify to remove `isValid` and not raise an exception if the trait is unsupported (can be checked via `Specification::hasTrait` so the functionality exists elsewhere), and instead add another `TraitPropertyStatus::kUnsupportedTrait`, then the workflow simplifies to
```c++
property::Int myValue = 123;  // Default if property not found.

if (MyTrait{specification}.getMyValue(&myValue) != TraitPropertyStatus::kFound) {

  std::cerr << "Warning: myValue not found\n";
}
```

[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/trait/TraitBase.hpp]: I guess this relates to the comment re. allowing mutators to set properties and implicitly make the trait supported, rather than requiring specifications to be pre-constructed with the supported traits.

[the remark this answers is not in the document]
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/trait/TraitBase.hpp]: Like the proposition, shall we do it in a follow up PR?

File: src/openassetio-core/include/openassetio/specification/Specification.hpp (hunk 19 above)
Comment (feltech) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:107]: Added some notes in c959d09f - I added a note to strongly recommend ASCII, which I'm not sure is really necessary these days, but would take some research to find out.

File: src/openassetio-core/include/openassetio/trait/TraitBase.hpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/trait/TraitBase.hpp]: > Restructured slightly in [bc85085](https://github.com/TheFoundryVisionmongers/OpenAssetIO/commit/bc8508552b00ee92ace36906d9a647aeabe0b012) and fixed the example.

This is def on the right track, and the example is great. What do you think about moving it above all the requirement for a derived class?

File: src/openassetio-core/include/openassetio/specification/Specification.hpp (hunk 19 above)
Comment (foundrytom) [reply src/openassetio-core/include/openassetio/specification/Specification.hpp:107]: GitHub won't let me resolve this comment 🤷‍♂️

File: src/openassetio-core/include/openassetio/trait/TraitBase.hpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (feltech) [reply src/openassetio-core/include/openassetio/trait/TraitBase.hpp]: Move up in bdb6c289 - seems to read well like that.
